bin/build.sh

```diff
@@ -335,90 +348,97 @@ then
 fi
 
 
-if [ ${BUILD_Z3} -eq 1 ]
-then
-  puts "Installing Z3"
-
-  ${WGET} ${Z3_DOWNLOAD_LINK} -O z3-downloaded.zip
-  unzip -o z3-downloaded.zip -d z3-extracted
-  mv -f --backup=numbered z3-extracted/z3-* ${Z3_DIR}
-  rm -rf z3-downloaded.zip z3-extracted
-
-  puts "Installed Z3"
+if [ ${BUILD_Z3} -eq 1 ] ; then
+  if [ ! -d "$Z3_DIR" ] ; then
+    puts "Installing Z3"
+    mkdir -p ${Z3_DIR}
+    ${WGET} ${Z3_DOWNLOAD_LINK} -O z3-downloaded.zip
+    unzip -o z3-downloaded.zip -d z3-extracted
+    mv -f --backup=numbered z3-extracted/z3-*/* ${Z3_DIR}
+    rm -rf z3-downloaded.zip z3-extracted
+    puts "Installed Z3"
+  else
+    puts "Z3 already built"
+  fi
 fi
 
 
-if [ ${BUILD_BOOGIE} -eq 1 ]
-then
-  puts "Building Boogie"
-
-  if [ ! -d "$BOOGIE_DIR" ] ; then
-    git clone https://github.com/boogie-org/boogie.git ${BOOGIE_DIR}
+if [ ${BUILD_BOOGIE} -eq 1 ] ; then
+  if [ "$(upToDate $BOOGIE_DIR $BOOGIE_COMMIT)" == "false" ] ; then
```

Comment:
Just nitpicking here, but why not use the standard exit-code condition here? e.g., where `upToDate` returns 0 on success, and the condition is
````
upToDate $BOOGIE_DIR $BOOGIE_COMMIT -eq 0
````

Reply:
I think my last commit fixed this. Good point.